arch/arm/include/kernel_arch_func.h

```diff
@@ -103,16 +98,32 @@ z_arch_switch_to_main_thread(struct k_thread *main_thread,
 #endif
 #endif /* CONFIG_BUILTIN_STACK_GUARD */
 
-	__set_PSP((u32_t)start_of_main_stack);
-	unlock_interrupts();
-#ifdef CONFIG_ARM_MPU
 	/*
-	 * If stack protection is enabled, make sure to set it
-	 * before jumping to thread entry function
+	 * Set PSP to the highest address of the main stack
+	 * before enabling interrupts and jumping to main.
 	 */
-	z_arch_configure_dynamic_mpu_regions(main_thread);
-#endif
-	z_thread_entry(_main, 0, 0, 0);
+	__asm__ volatile (
+	"mov   r4,  %0     \n\t"   /* Store _main in R4 */
```

Comment:
Any reason why you use r4 here instead of storing it in r0 directly and using any other registr to set BASEPRI later?

Reply:
Hm, right :)

Well, I thought this is a bit more robust against "someone" going there and inserting a "bl <some routine> after setting PSP and before branching to z_thread_entry. E.g. something related to tracing, bench-marking, profiling, etc.

Reply:
@carlescufi, modified patch to use R0 (and R1 for BASEPRI), based on your suggestion